src/fast_scanning.jl

```diff
@@ -0,0 +1,235 @@
+            result[point] = new_label
+            region_means[new_label] = img[point]
+            region_pix_count[new_label] = 1
+            push!(labels, new_label)
+
+        # Else if one valid label found, assign that label
+        elseif top_label == -1 || left_label == -1
+            valid_label = top_label == -1 ? left_label : top_label
+            result[point] = valid_label
+            region_pix_count[valid_label] += 1
+            region_means[valid_label] += (img[point] - region_means[valid_label])/(region_pix_count[valid_label])
+
+        # Else if both labels are valid
+        else
+            # If both the labels are same, assign that label
+            if top_label == left_label
+                result[point] = top_label
+                region_pix_count[top_label] += 1
+                region_means[top_label] += (img[point] - region_means[top_label])/(region_pix_count[top_label])
+
+            # Else replace label having less pixel count with that having larger one and assign the label having larger pixel count
+            else
+                min_count_point = region_pix_count[top_label] < region_pix_count[left_label] ? top_point: left_point
+                min_count_label = result[min_count_point]
+                max_count_label = min_count_label == top_label ? left_label : top_label
+
+                result[point] = max_count_label
+                bfs_color!(result, min_count_point, max_count_label)
+
+                region_pix_count[min_count_label] += 1
+                region_pix_count[max_count_label] += region_pix_count[min_count_label]
+                region_means[min_count_label] += (img[point] - region_means[min_count_label])/(region_pix_count[min_count_label])
+                region_means[max_count_label] += (region_means[min_count_label] - region_means[max_count_label])*region_pix_count[min_count_label]/region_pix_count[max_count_label]
+
+                # Remove label result[i] from region_means, region_pix_count
+                delete!(region_pix_count, min_count_label)
+                delete!(region_means, min_count_label)
+                push!(removed_labels, min_count_label)
+            end
+        end
+    end
+
+    # Remove labels that have been recolored due to the bfs calls
+    @noinline sort_fn!(removed_labels) = sort!(removed_labels)
+    sort_fn!(removed_labels)
+    deleteat!(labels, removed_labels)
+
+    SegmentedImage(result, labels, region_means, region_pix_count)
+
+end
+
+
+# For N-D images
+
+function fast_scanning{CT<:Colorant,N}(img::AbstractArray{CT,N}, threshold::Real, diff_fn::Function = default_diff_fn)
+
+    # Neighbourhood functions
+    _diagmN = diagm([1 for i in 1:N])
+    half_region = ntuple(i-> CartesianIndex{N}(ntuple(j->_diagmN[j,i], Val{N})), Val{N})
+    neg_neighbourhood{N}(x::CartesianIndex{N}) = ntuple(i-> x-half_region[i], Val{N})
+    pos_neighbourhood{N}(x::CartesianIndex{N}) = ntuple(i-> x+half_region[i], Val{N})
```

Comment:
So I guess one difference between the two implementations is that "corners" are used here but not above. I wonder if one could unify the two algorithms by having a `AbstractStencilIterator`, with different concrete implementations, and allowing that as an argument to the function?